inference-engine/src/mkldnn_plugin/nodes/non_max_suppression.cpp

```diff
@@ -135,110 +241,196 @@ class NonMaxSuppressionImpl: public ExtLayerBase {
-
-        for (int batch = 0; batch < num_batches; batch++) {
-            float *boxesPtr = boxes + batch * boxesStrides[0];
-            for (int class_idx = 0; class_idx < num_classes; class_idx++) {
-                float *scoresPtr = scores + batch * scoresStrides[0] + class_idx * scoresStrides[1];
-                std::vector<std::pair<float, int> > scores_vector;
-                for (int box_idx = 0; box_idx < num_boxes; box_idx++) {
-                    if (scoresPtr[box_idx] > score_threshold)
-                        scores_vector.push_back(std::make_pair(scoresPtr[box_idx], box_idx));
-                }
+        const int &num_batches = static_cast<int>(scores_dims[0]);
+        const int &num_classes = static_cast<int>(scores_dims[1]);
+
+        std::vector<filteredBoxes> filtBoxes(max_output_boxes_per_class * num_batches * num_classes);
+
+        auto coeff = [&](float iou) {
+            const float weight = std::exp(scale * iou * iou);
+            return iou <= iou_threshold ? weight : 0.0f;
+        };
+
+        auto less = [](const boxInfo& l, const boxInfo& r) {
+            return l.score < r.score || ((l.score == r.score) && (l.idx > r.idx));
+        };
 
-                if (scores_vector.size()) {
-                    parallel_sort(scores_vector.begin(), scores_vector.end(),
-                        [](const std::pair<float, int>& l, const std::pair<float, int>& r) { return l.first > r.first; });
-
-                    int io_selection_size = 1;
-                    fb.push_back({ scores_vector[0].first, batch, class_idx, scores_vector[0].second });
-                    for (int box_idx = 1; (box_idx < static_cast<int>(scores_vector.size()) && io_selection_size < max_output_boxes_per_class); box_idx++) {
-                        bool box_is_selected = true;
-                        for (int idx = io_selection_size - 1; idx >= 0; idx--) {
-                            float iou = intersectionOverUnion(&boxesPtr[scores_vector[box_idx].second * 4],
-                                             &boxesPtr[scores_vector[idx].second * 4], center_point_box);
-                            if (iou > iou_threshold) {
-                                box_is_selected = false;
-                                break;
-                            }
+        parallel_for2d(num_batches, num_classes, [&](int batch_idx, int class_idx) {
+            std::vector<filteredBoxes> fb;
+            const float *boxesPtr = boxes + batch_idx * boxesStrides[0];
+            const float *scoresPtr = scores + batch_idx * scoresStrides[0] + class_idx * scoresStrides[1];
+
+            std::priority_queue<boxInfo, std::vector<boxInfo>, decltype(less)> sorted_boxes(less);
+            for (int box_idx = 0; box_idx < num_boxes; box_idx++) {
+                if (scoresPtr[box_idx] > score_threshold)
+                    sorted_boxes.emplace(boxInfo({scoresPtr[box_idx], box_idx, 0}));
+            }
+
+            if (sorted_boxes.size() > 0) {
```

Comment:
Seems, we can add fb.reserve(max_output_boxes_per_class) here.

Reply:
added